File: app/src/main/java/eu/kanade/mangafeed/ui/setting/SettingsAccountsFragment.java
package eu.kanade.mangafeed.ui.setting;

import android.os.Bundle;
import android.preference.PreferenceCategory;
import android.preference.PreferenceScreen;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import javax.inject.Inject;

import eu.kanade.mangafeed.App;
import eu.kanade.mangafeed.data.mangasync.base.MangaSyncService;
import eu.kanade.mangafeed.data.mangasync.MangaSyncManager;
import eu.kanade.mangafeed.data.source.SourceManager;
import eu.kanade.mangafeed.data.source.base.Source;
import eu.kanade.mangafeed.ui.setting.preference.MangaSyncLoginDialog;
import eu.kanade.mangafeed.ui.setting.preference.SourceLoginDialog;
import rx.Observable;

public class SettingsAccountsFragment extends SettingsNestedFragment {

    @Inject SourceManager sourceManager;
    @Inject MangaSyncManager syncManager;

    public static SettingsNestedFragment newInstance(int resourcePreference, int resourceTitle) {
        SettingsNestedFragment fragment = new SettingsAccountsFragment();
        fragment.setArgs(resourcePreference, resourceTitle);
        return fragment;
    }

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        App.get(getActivity()).getComponent().inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedSate) {
        View view = super.onCreateView(inflater, container, savedSate);

        PreferenceScreen screen = getPreferenceScreen();

        List<Source> sourceAccounts = getSourcesWithLogin();

        PreferenceCategory sourceCategory = new PreferenceCategory(screen.getContext());
        sourceCategory.setTitle("Sources");
        screen.addPreference(sourceCategory);

        for (Source source : sourceAccounts) {
            SourceLoginDialog dialog = new SourceLoginDialog(
                    screen.getContext(), preferences, source);
            dialog.setTitle(source.getName());

            sourceCategory.addPreference(dialog);
        }

        PreferenceCategory mangaSyncCategory = new PreferenceCategory(screen.getContext());
        mangaSyncCategory.setTitle("Sync");
        screen.addPreference(mangaSyncCategory);

        for (MangaSyncService sync : syncManager.getSyncServices()) {
            MangaSyncLoginDialog dialog = new MangaSyncLoginDialog(
                    screen.getContext(), preferences, sync);
            dialog.setTitle(sync.getName());

            mangaSyncCategory.addPreference(dialog);
        }

        return view;
    }

    private List<Source> getSourcesWithLogin() {
        return Observable.from(sourceManager.getSources())
                .filter(Source::isLoginRequired)
                .toList()
                .toBlocking()
                .single();
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/ui/setting/SettingsAccountsFragment.kt
package eu.kanade.tachiyomi.ui.setting

import android.content.Context
import android.os.Bundle
import android.support.v7.preference.DialogPreference
import android.support.v7.preference.Preference
import android.support.v7.preference.PreferenceCategory
import android.view.View
import eu.kanade.tachiyomi.R
import eu.kanade.tachiyomi.data.source.base.Source
import eu.kanade.tachiyomi.widget.preference.MangaSyncLoginDialog
import eu.kanade.tachiyomi.widget.preference.SourceLoginDialog

class SettingsAccountsFragment : SettingsNestedFragment() {

    companion object {

        fun newInstance(resourcePreference: Int, resourceTitle: Int): SettingsNestedFragment {
            val fragment = SettingsAccountsFragment()
            fragment.setArgs(resourcePreference, resourceTitle)
            return fragment
        }
    }

    val sourceCategory by lazy { findPreference("pref_category_source_accounts") as PreferenceCategory }
    val syncCategory by lazy { findPreference("pref_category_manga_sync_accounts") as PreferenceCategory }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val themedContext = preferenceManager.context

        for (source in getSourcesWithLogin()) {
            val pref = SourcePreference(themedContext).apply {
                isPersistent = false
                title = source.name
                key = source.id.toString()
                dialogLayoutResource = R.layout.pref_account_login
            }

            sourceCategory.addPreference(pref)
        }

        for (sync in settingsActivity.syncManager.services) {
            val pref = SyncPreference(themedContext).apply {
                isPersistent = false
                title = sync.name
                key = sync.id.toString()
                dialogLayoutResource = R.layout.pref_account_login
            }

            syncCategory.addPreference(pref)
        }
    }

    fun getSourcesWithLogin(): List<Source> {
        return settingsActivity.sourceManager.sources.filter { it.isLoginRequired }
    }

    override fun onDisplayPreferenceDialog(preference: Preference) {
        if (preference is SourcePreference) {
            val fragment = SourceLoginDialog.newInstance(preference)
            fragment.setTargetFragment(this, 0)
            fragment.show(childFragmentManager, null)
        } else if (preference is SyncPreference) {
            val fragment = MangaSyncLoginDialog.newInstance(preference)
            fragment.setTargetFragment(this, 0)
            fragment.show(childFragmentManager, null)
        } else {
            super.onDisplayPreferenceDialog(preference)
        }
    }

    class SourcePreference(context: Context) : DialogPreference(context) {}

    class SyncPreference(context: Context) : DialogPreference(context) {}

}
